### PR TITLE
Feature/ui rework

### PR DESCRIPTION
**Pull request by friedow**

### CODE REVIEW

File: src/tetris/scene/playScene.ts [hunk 1]
```diff
@@ -11,8 +11,65 @@ import Vector2 = Phaser.Math.Vector2;
...
-const PLAYER_FIELD_DRAW_OFFSET: Vector2 = new Vector2(50, 80);
+import CountdownWidget from "tetris/ui/countdownWidget";
+import ScoreboardWidget from "tetris/ui/scoreboardWidget";
+import ScoreWidget from "tetris/ui/scoreWidget";
+
+const PLAYER_FIELD_DRAW_OFFSET: Vector2 = new Vector2(
+	(config.graphics.width - config.field.width * config.field.blockSize) / 2, 
+	(config.graphics.height - config.field.height * config.field.blockSize) / 1.5);
+
+const players = [
```
Comment (julianweise) [inline src/tetris/scene/playScene.ts]: Shouldn't we remove this static array? Or is it for illustrating purpose and gets removed with a future PR?

File: src/tetris/scene/playScene.ts [hunk 2]
```diff
@@ -138,11 +205,36 @@ export default class PlayScene extends Phaser.Scene {
 	}
 
 	private _createUi() {
```
Comment (julianweise) [inline src/tetris/scene/playScene.ts]: missing `:void`

File: src/tetris/scene/playScene.ts [hunk 3]
```diff
@@ -138,11 +205,36 @@ export default class PlayScene extends Phaser.Scene {
...
-		this._scoreText = this.add.text(PLAYER_FIELD_DRAW_OFFSET.x, PLAYER_FIELD_DRAW_OFFSET.y, "0", config.defaultLargeFontStyle);
+
+		// create UI widgets
+		this._scoreWidget = new ScoreWidget(this, config.graphics.width / 2, (config.graphics.height - config.field.height * config.field.blockSize) / 4);
+		this._countdownWidget = new CountdownWidget(this, config.graphics.width / 5 * 4, config.graphics.height / 30 * 8);
+		this._scoreboardWidget = new ScoreboardWidget(this, config.graphics.width / 5 * 4, config.graphics.height / 20 * 9);
+		this._scoreboardWidget.update(players);
+	}
+
+	private _initializeShaders() {
```
Comment (julianweise) [inline src/tetris/scene/playScene.ts]: missing `:void`

File: src/tetris/scene/playScene.ts [hunk 4]
```diff
@@ -138,11 +205,36 @@ export default class PlayScene extends Phaser.Scene {
...
+		this._pipeline = new Phaser.Renderer.WebGL.Pipelines.TextureTintPipeline({
+			game: this._game,
+			renderer: this._game.renderer, 
+			fragShader: this.cache.shader.get('rainbow') 
+		});
+		(this._game.renderer as Phaser.Renderer.WebGL.WebGLRenderer).addPipeline('rainbow-text', this._pipeline);
+		this._pipeline.setFloat2('uResolution', config.graphics.width, config.graphics.height);
+	}
+
+	private _updateShaders(time: number) {
```
Comment (julianweise) [inline src/tetris/scene/playScene.ts]: missing `:void`

File: src/tetris/ui/countdownWidget.ts [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+export default class CountdownWidget {
+
+	//region public members
+	public text: Phaser.GameObjects.Text;
+	public circle: Phaser.GameObjects.Graphics;
+	//endregion
+
+	//region public methods
+	get height() {
```
Comment (julianweise) [inline src/tetris/ui/countdownWidget.ts]: Return type missing

File: src/tetris/ui/countdownWidget.ts [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+		return this._radius * 2;
+	}
+
+	get width() {
```
Comment (julianweise) [inline src/tetris/ui/countdownWidget.ts]: Return type missing

File: src/tetris/ui/countdownWidget.ts [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+		return this._radius * 2;
+	}
+
+	get x() {
```
Comment (julianweise) [inline src/tetris/ui/countdownWidget.ts]: Return type missing

File: src/tetris/ui/countdownWidget.ts [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+		return this.circle.x;
+	}
+
+	get y() {
```
Comment (julianweise) [inline src/tetris/ui/countdownWidget.ts]: Return type missing

File: src/tetris/ui/countdownWidget.ts [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+		this.circle.x = x;
+		this._adjustTextPosition();
+	}
+
+	set y(y: number) {
+		this.circle.y = y;
+		this._adjustTextPosition();
+	}
+
+	public update(time: number, totalTime: number) {
```
Comment (julianweise) [inline src/tetris/ui/countdownWidget.ts]: missing `:void`

File: src/tetris/ui/countdownWidget.ts [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+		this.update(30, 30);
+	}
+	//endregion
+
+	//region private members
+	private readonly _radius = config.graphics.width / 15;
+	//endregion
+
+	//region private methods
+	private _updateCircle(percentage: number) {
```
Comment (julianweise) [inline src/tetris/ui/countdownWidget.ts]: missing `:void`

File: src/tetris/ui/countdownWidget.ts [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+			this.circle.arc(0, 0, this._radius, startRad, Phaser.Math.DegToRad(269), false);
+			this.circle.closePath();
+		} else {
+			this.circle.arc(0, 0, this._radius, startRad, endRad, false);
+		}
+		
+		this.circle.strokePath();
+	}
+
+	private _updateText(percentage: number, time: number) {
```
Comment (julianweise) [inline src/tetris/ui/countdownWidget.ts]: missing `:void`

File: src/tetris/ui/countdownWidget.ts [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+		if (percentage < 25) {
+			this.text.setText(time.toFixed(3).toString());
+		} else {
+			this.text.setText(time.toFixed(0).toString());
+		}
+		this._adjustTextPosition();
+	}
+
+	private _setColors(percentage: number) {
```
Comment (julianweise) [inline src/tetris/ui/countdownWidget.ts]: missing `:void`

File: src/tetris/ui/countdownWidget.ts [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+		} else if (percentage < 75) {
+			this.circle.lineStyle(config.ui.countdown.lineWidth, config.ui.colors.yellow.hex);
+			this.text.setColor(config.ui.colors.yellow.string);
+		} else {
+			this.circle.lineStyle(config.ui.countdown.lineWidth, config.ui.colors.white.hex);
+			this.text.setColor(config.ui.colors.white.string);
+		}
+	}
+	
+	private _adjustTextPosition() {
```
Comment (julianweise) [inline src/tetris/ui/countdownWidget.ts]: missing `:void`

File: src/tetris/ui/countdownWidget.ts [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+		this.text.x = this.x - this.text.width / 2;
+		this.text.y = this.y - this.text.height / 1.5;
+	}
+
+	private static _phaserRadius(percentage: number) {
```
Comment (julianweise) [inline src/tetris/ui/countdownWidget.ts]: missing `:void`

File: src/tetris/ui/scoreWidget.ts [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+import { Font } from "tetris/ui/scoreboardWidget";
+
+export default class ScoreWidget {
+
+	//region public members
+	public score: Phaser.GameObjects.Text;
+	//endregion
+
+	//region public methods
+	get height() {
```
Comment (julianweise) [inline src/tetris/ui/scoreWidget.ts]: All getter are missing proper return type information

File: src/tetris/ui/scoreWidget.ts [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+        this._x = x;
+        this._adjustTextX();
+	}
+
+	set y(y: number) {
+        this._y = y;
+		this._adjustTextY();
+	}
+
+	public update(score: string) {
```
Comment (julianweise) [inline src/tetris/ui/scoreWidget.ts]: missing `:void`

File: src/tetris/ui/scoreWidget.ts [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+	}
+	//endregion
+
+	//region private members
+    private _x: number = 0;
+    private _y: number = 0;
+	//endregion
+
+    //region private methods
+    private _adjustTextX() {
```
Comment (julianweise) [inline src/tetris/ui/scoreWidget.ts]: missing `:void`

File: src/tetris/ui/scoreWidget.ts [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+        this.score.x = this._x - this.score.width / 2;
+    }
+
+    private _adjustTextY() {
```
Comment (julianweise) [inline src/tetris/ui/scoreWidget.ts]: missing `:void`

File: src/tetris/ui/scoreboardWidget.ts [hunk 19]
```diff
@@ -0,0 +1,166 @@
...
+            this.names[index].y += offset;
+            this.scores[index].y += offset;
+        }
+        for (const divider of this.dividers) {
+            divider.y += offset;
+        }
+        this._y = y;
+	}
+
+	public update(playerScores: PlayerScore[]) {
```
Comment (julianweise) [inline src/tetris/ui/scoreboardWidget.ts]: missing `:void`

File: src/tetris/ui/scoreboardWidget.ts [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+
+	//region private members
+    private _scene: Phaser.Scene
+    private _x: number = 0;
+    private _y: number = 0;
+    private _font: Font;
+	//endregion
+
+    //region private methods
+    private _clearScoreboard() {
```
Comment (julianweise) [inline src/tetris/ui/scoreboardWidget.ts]: missing `:void`

File: src/tetris/ui/scoreboardWidget.ts [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+            this.ranks[index].destroy();
+            this.names[index].destroy();
+            this.scores[index].destroy();
+        }
+        this.ranks.splice(0, this.ranks.length - 1);
+        this.names.splice(0, this.names.length - 1);
+        this.scores.splice(0, this.scores.length - 1);
+    }
+
+    private _createPlayerText(index: number, player: PlayerScore) {
```
Comment (julianweise) [inline src/tetris/ui/scoreboardWidget.ts]: missing `:void`

File: src/tetris/ui/scoreboardWidget.ts [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+            name.setPipeline('rainbow-text');
+            score.setPipeline('rainbow-text');
+        } else if (player.ownScore === true) {
+            rank.setColor(config.ui.colors.yellow.string);
+            name.setColor(config.ui.colors.yellow.string);
+            score.setColor(config.ui.colors.yellow.string);
+        }
+    }
+
+    private _createDivider(index: number, color: number, double: boolean = false) {
```
Comment (julianweise) [inline src/tetris/ui/scoreboardWidget.ts]: missing `:void`

File: src/tetris/ui/scoreboardWidget.ts [hunk 23]
```diff
@@ -0,0 +1,166 @@
...
+        name.y = this.y + index * (config.ui.spacing + this._font.size);
+
+        // Add score text
+        const score = this._scene.add.text(0, 0, player.score, this._font.font);
+        score.x = this.x + 125 - score.width / 2;
+        score.y = this.y + index * (config.ui.spacing + this._font.size);
+
+        // Set text color
+        if (player.danger) {
+            rank.setColor(config.ui.colors.red.string);
```
Comment (julianweise) [inline src/tetris/ui/scoreboardWidget.ts]: 😢 So sad that we can't use css classes inside canvas ...

File: src/tetris/scene/playScene.ts (hunk 1 above)
Comment (friedow) [reply src/tetris/scene/playScene.ts]: Yes, it's just for illustration purposes and will be removed later.

**Comment by julianweise** [comment]
Added missing return types, added missing terminate statements and simplified one condition in order to get this PR mergeable (blocks current development)

**Pull request merged**